README.md
## Banner Listtile

[<img src="https://img.shields.io/badge/pub-v2.3.0-blue.svg">](https://pub.dev/packages/banner_listtile)
[<img src="https://img.shields.io/badge/License-MIT-green.svg">](https://pub.dev/packages/banner_listtile/license)
<br />
Banner_Listtile comes in handy when you need a temple of tile with a built-in banner option. You may have seen a banner with a tag ex. "New," "Hot," "New Collection" etc. Banner_listtile does the exact thing, and also gives you the extra feature of customizing it easily.


## Installation
i. Add the latest version of this package to your pubspec.yaml file, and run: 'dart pub get':
```yaml
dependancies:
  banner_listtile: ^2.3.0
```

ii. Import the package in your Flutter App and use it.
```dart
import 'package:banner_listtile/banner_listtile.dart';
```

## Changes
<table>
<tr><td  valign="top">

## New 2.3.0 Changes <br />
- ✅ Position enum added. Instead of boolean now Position enum required as its value. <br />
- 🔀 Now Icons can be added on banner in place of text.
- 🚫 RandomColor property removed from this package.
- ✨ Fixed UI bleeding issue as now zigzagimage side looks smoother and sharp.
</td>



</tr>
</table>

## Note:
If you use it inside a Column, Listview or any kind of vertical list, then you do not need to give it a height property. But when you use it as a single widget or inside any other widget, then you are suggested to define its height by giving it a height property. If you don't, then it might end up taking all the available space.

## Sample Code

```dart
class BannerListTileExample extends StatelessWidget {
  const BannerListTileExample({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: BannerListTile(
          backgroundColor: Colors.blue,
          borderRadius: BorderRadius.circular(8),
          bannerPosition: BannerPosition.topRight,
          imageContainer: Image(
              image: NetworkImage(
                  "https://images.unsplash.com/photo-1529626455594-4ff0802cfb7e?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb1.2.1&auto=format&fit=crop&w=387&q=80"),
              fit: BoxFit.cover),
          title: Text(
            "Lisa",
            style: TextStyle(fontSize: 24, color: Colors.white),
            overflow: TextOverflow.ellipsis,
            maxLines: 1,
          ),
          subtitle: Text("A model from NY",
              style: TextStyle(fontSize: 13, color: Colors.white)),
          trailing: IconButton(
              onPressed: () {},
              icon: Icon(
                Icons.delete_forever,
                color: Colors.red,
              )),
        ),
      ),
    );
  }
}
```
<img  src="https://user-images.githubusercontent.com/60544576/140290340-e166477e-7113-4032-8546-e51cc8e655ed.png" width="400" alt="">

## More example: (Add text or icon)
<img  src="https://user-images.githubusercontent.com/60544576/236819021-74060f0e-0e71-4152-bf68-f9846dd43295.png"  alt="Sample Image">


## New zigzag shape
```flutter
BannerListTile(
    imageContainerShapeZigzagIndex: index, // [index = (even or odd integer number)]
),
```
<img  src="https://user-images.githubusercontent.com/60544576/140294624-de6c38c2-21a7-4e09-80f8-786173b2a51c.png"  alt="">


## Examples
<img src="https://user-images.githubusercontent.com/60544576/236823584-9e2be81b-8923-447c-9930-d5486bb32af1.png" width="400" alt="example screenshot">





## Upcoming

Things are added & also will be added more if needed in future.
Contributions to this package is welcomed. For this go to github repository of this package and fork this repository.

If you have any suggestions or bug report then please consider mailing me at:

```dart 
eaglex129@gmail[dot]com
```


CHANGELOG.md
## 2.3.0
- ✅ Position enum added. Instead of boolean, now it takes a Position enum.
- 🔀 Now Icons can be added on banner in place of text.
- 🚫 RandomColor property removed from this package.
- ✨ Fixed UI bleeding issue as now zigzagimage side looks smoother and sharp.

## 2.2.2
- Documentation updated.

## 2.2.1
- All dart plugins are upgraded and up to date (Flutter 3.7).
- Readme updated.

## 2.2.0
- Added functionalities. Example: onTap, onDoubleTap, & more.

## 2.1.0
- Added elevation/shadow.
- Added option to give side border.
- Now margin can be added.

## 2.0.0
- Fixed few bugs.
- Dynamically height change by Title or Subtitle data.
- Added Subtitle opacity.
- Now user can change trailing width.
- User now give custom size to banner.
- Trailing no longer be visible without its data.
- Added a imagecontainer index option to show zigzag shape in list.
- Now imagecontainer size can be changed.
- Trailing now can be shown at the center or top. (Size depends on imagecontainer size)

## 1.0.0
- Added banner
- Leading image option with custom shape container
- Rounded corner container option
- Title,subtitle, & trailing button
- & few customization like color & few others
- Random color option for the tile

test/banner_listtile_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:banner_listtile/banner_listtile.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


example/lib/main.dart

import 'package:flutter/material.dart';

import 'custom_screen.dart';
import 'listview_screen.dart';

void main() {
  runApp( const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      //debugShowCheckedModeBanner: false,
      title: 'Flutter Demo',
      theme: ThemeData.light(),
      home:  const ContainerTestView(),
    );
  }
}

class ContainerTestView extends StatefulWidget {
   const ContainerTestView({Key? key}) : super(key: key);



  @override
  State<ContainerTestView> createState() => _ContainerTestViewState();
}

class _ContainerTestViewState extends State<ContainerTestView> {
  int _selectedIndex = 0;

  static final List<Widget> _pages = <Widget>[
    const ListviewScreen(),
    const CustomScreen(),

  ];

   void _onItemTapped(int index) {
     setState(() {
       _selectedIndex = index;
     });
   }

  @override
  Widget build(BuildContext context) {

    return Scaffold(
        backgroundColor: Colors.white,
        appBar: AppBar(
          title: const Text("Banner Listtile"),
        ),
      body: _pages.elementAt(_selectedIndex),
      bottomNavigationBar: BottomNavigationBar(
        currentIndex: _selectedIndex, //New
        onTap: _onItemTapped,
        items: const <BottomNavigationBarItem>[
          BottomNavigationBarItem(
            icon: Icon(Icons.menu),
            label: 'List',
          ),
          BottomNavigationBarItem(
            icon: Icon(Icons.dashboard_customize),
            label: 'Custom',
          ),

        ],
      ),
    );
  }
}


example/lib/listview_screen.dart
import 'package:banner_listtile/banner_listtile.dart';
import 'package:flutter/material.dart';

class ListviewScreen extends StatelessWidget {
  const ListviewScreen({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return ListView.builder(itemCount: 6,itemBuilder: (ctx , index){
      return Padding(
        padding: const EdgeInsets.all(5.0),
        child: BannerListTile(

          onTap: (){},
          bannerText: "banner",


          imageContainerShapeZigzagIndex: index,
          bannerPosition: BannerPosition.topLeft,
          borderRadius: BorderRadius.circular(8),
          imageContainer: const Image(
              image: NetworkImage(
                  "https://images.unsplash.com/photo-1529626455594-4ff0802cfb7e?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb1.2.1&auto=format&fit=crop&w=387&q=80"),
              fit: BoxFit.cover),
          title: const Text(
            "Lisa",
            style: TextStyle(fontSize: 24, color: Colors.white),
            overflow: TextOverflow.ellipsis,
            maxLines: 1,
          ),
          subtitle: const Text("A model from NY",
              style: TextStyle(fontSize: 13, color: Colors.white)),
          trailing: IconButton(
              onPressed: () {},
              icon: const Icon(
                Icons.delete_forever,
                color: Colors.red,
              )),
        ),
      );
    });
  }
}


example/lib/custom_screen.dart
import 'package:banner_listtile/banner_listtile.dart';
import 'package:flutter/material.dart';

class CustomScreen extends StatelessWidget {
  const CustomScreen({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.only(top:8.0,right: 8.0,left:8.0),
      child: SingleChildScrollView(
        child: Column(
          children: [
            BannerListTile(
              onTap: (){},
              backgroundColor: Colors.blue,
              borderRadius: BorderRadius.circular(8),

              imageContainer: const Image(
                  image: NetworkImage(
                      "https://images.unsplash.com/photo-1529626455594-4ff0802cfb7e?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb1.2.1&auto=format&fit=crop&w=387&q=80"),
                  fit: BoxFit.cover),

              title: const Text(
                "Monalisa",
                style: TextStyle(fontSize: 24, color: Colors.white),
                overflow: TextOverflow.ellipsis,
                maxLines: 1,
              ),
              subtitle: Row(
                mainAxisAlignment: MainAxisAlignment.start,
                crossAxisAlignment: CrossAxisAlignment.center,
                children: [
                  Row(
                    children: [
                      IconButton(
                          onPressed: () {},
                          icon: const Icon(
                            Icons.facebook_outlined,
                            color: Colors.black87,
                          )),
                      IconButton(
                          onPressed: () {},
                          icon: const Icon(
                            Icons.location_on_outlined,
                            color: Colors.black87,
                          )),
                      IconButton(
                          onPressed: () {},
                          icon: const Icon(
                            Icons.phone,
                            color: Colors.black87,
                          )),
                    ],
                  ),
                ],
              ),
              trailing: IconButton(
                  onPressed: () {},
                  icon: const Icon(
                    Icons.cancel,
                    color: Colors.red,
                  )),
            ),
            const SizedBox(height: 4),
            //
            BannerListTile(
              onTap: (){},
              elevation: 8,
              bannerPosition: BannerPosition.topRight,
              bannerColor: Colors.amber,
              bannerTextColor: Colors.red,
              bannerText: "Hot",
              bannerIcon: const Icon(Icons.new_releases_outlined),
              //bannerIcon: Container(color: Colors.green,height: 200,width: 200),
              bannerIconRotation: 1,
              bannerSize: 45,

              borderRadius: BorderRadius.circular(8),
              imageContainer: const Image(
                  image: NetworkImage(
                      "https://images.unsplash.com/photo-1529626455594-4ff0802cfb7e?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb1.2.1&auto=format&fit=crop&w=387&q=80"),
                  fit: BoxFit.cover),
              title: const Text(
                "Lisa",
                style: TextStyle(fontSize: 24, color: Colors.white),
                overflow: TextOverflow.ellipsis,
                maxLines: 1,
              ),
              subtitle: const Text("A model from NY",
                  style: TextStyle(fontSize: 13, color: Colors.white)),
              trailing: IconButton(
                  onPressed: () {},
                  icon: const Icon(
                    Icons.delete_forever,
                    color: Colors.red,
                  )),
            ),
            //
            const SizedBox(height: 4),
            BannerListTile(
              onTap: (){},
              bannerPosition: BannerPosition.topLeft,
              bannerColor: Colors.amber,
              bannerText: "New Item",
              bannerTextColor: Colors.red,
              backgroundColor: Colors.blue,
              borderRadius: BorderRadius.circular(8),
              imageContainer: const Image(
                  image: NetworkImage(
                      "https://images.unsplash.com/photo-1529626455594-4ff0802cfb7e?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb1.2.1&auto=format&fit=crop&w=387&q=80"),
                  fit: BoxFit.cover),
              title: const Text(
                "Monalisa",
                style: TextStyle(fontSize: 24, color: Colors.white),
                overflow: TextOverflow.ellipsis,
                maxLines: 1,
              ),
              subtitle: Row(
                mainAxisAlignment: MainAxisAlignment.start,
                crossAxisAlignment: CrossAxisAlignment.center,
                children: [
                  Row(
                    children: [
                      IconButton(
                          onPressed: () {},
                          icon: const Icon(
                            Icons.facebook_outlined,
                            color: Colors.black87,
                          )),
                      IconButton(
                          onPressed: () {},
                          icon: const Icon(
                            Icons.location_on_outlined,
                            color: Colors.black87,
                          )),
                      IconButton(
                          onPressed: () {},
                          icon: const Icon(

                            Icons.phone,
                            color: Colors.black87,
                          )),
                    ],
                  ),
                ],
              ),
              trailing: IconButton(
                  onPressed: () {},
                  icon: const Icon(
                    Icons.cancel,
                    color: Colors.red,
                  )),
            ),
            const SizedBox(height: 4),
            //
            // BannerListTile(
            //   onTap: (){},
            //   bannerText: "banner",
            //
            //
            //   bannerPosition: BannerPosition.topLeft,
            //   borderRadius: BorderRadius.circular(8),
            //   imageContainer: const Image(
            //       image: NetworkImage(
            //           "https://images.unsplash.com/photo-1529626455594-4ff0802cfb7e?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb1.2.1&auto=format&fit=crop&w=387&q=80"),
            //       fit: BoxFit.cover),
            //   title: const Text(
            //     "Lisa",
            //     style: TextStyle(fontSize: 24, color: Colors.white),
            //     overflow: TextOverflow.ellipsis,
            //     maxLines: 1,
            //   ),
            //   subtitle: const Text("A model from NY",
            //       style: TextStyle(fontSize: 13, color: Colors.white)),
            //   trailing: IconButton(
            //       onPressed: () {},
            //       icon: const Icon(
            //         Icons.delete_forever,
            //         color: Colors.red,
            //       )),
            // ),
            // //
            // const SizedBox(height: 4),
            //
            BannerListTile(
              onTap: (){},
              onLongPress: (){},

              showBanner: true,
              bannerText: "Big banner",
              subtitleOpacity: 1.0,

              bannerSize: 60,
              bannerPosition: BannerPosition.topRight,
              borderRadius: BorderRadius.circular(8),
              imageContainer: const Image(
                  image: NetworkImage(
                      "https://images.unsplash.com/photo-1529626455594-4ff0802cfb7e?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb1.2.1&auto=format&fit=crop&w=387&q=80"),
                  fit: BoxFit.cover),
              title: const Text(
                "Dummy Long Text",
                style: TextStyle(fontSize: 24, color: Colors.white),
                overflow: TextOverflow.ellipsis,
                maxLines: 1,
              ),
              subtitle: const Text("Lorem Ipsum is simply dummy text of the printing and typesetting industry. \nLorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. \n\nIt has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.",
                  style: TextStyle(fontSize: 13, color: Colors.white)),

            ),
          ],
        ),
      ),
    );
  }
}
